cmd/src/actions_exec_backend_runner.go

```diff
@@ -0,0 +1,307 @@
+		logWriter = io.MultiWriter(logWriter, os.Stderr)
+	}
+
+	x.updateRepoStatus(repo, ActionRepoStatus{
+		LogFile:   logFile.Name(),
+		StartedAt: time.Now(),
+	})
+
+	// TODO!(sqs)
+	//
+	// ctx, cancel := context.WithTimeout(ctx, 90*time.Second)
+	// defer cancel()
+
+	patch, err := runAction(ctx, prefix, repo.ID, repo.Name, repo.Rev, x.action.Run, logWriter)
+	status := ActionRepoStatus{
+		FinishedAt: time.Now(),
+	}
+	if len(patch) > 0 {
+		status.Patch = CampaignPlanPatch{
+			Repository:   repo.ID,
+			BaseRevision: repo.Rev,
+			Patch:        string(patch),
+		}
+	}
+	if err != nil {
+		status.Err = err
+		fmt.Fprintf(logWriter, "# ERROR: %s\n", err)
+	}
+	x.updateRepoStatus(repo, status)
+
+	// Add to cache if successful.
+	if err == nil {
+		if err := x.opt.cache.set(ctx, cacheKey, status.Patch); err != nil {
+			return errors.Wrapf(err, "caching result for %s", repo.Name)
+		}
+	}
+
+	return err
+}
+
+func runAction(ctx context.Context, prefix, repoID, repoName, rev string, runs []*ActionFileRun, logFile io.Writer) ([]byte, error) {
+	fmt.Fprintf(logFile, "# Repository %s @ %s (%d steps)\n", repoName, rev, len(runs))
+
+	zipFile, err := fetchRepositoryArchive(ctx, repoName, rev)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(zipFile.Name())
+
+	volumeDir, err := unzipToTempDir(ctx, zipFile.Name(), prefix)
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(volumeDir)
+
+	for i, run := range runs {
+		if i != 0 {
+			fmt.Fprintln(logFile)
+		}
+
+		logPrefix := fmt.Sprintf("Step %d", i)
+
+		switch run.Type {
+		case "command":
+			fmt.Fprintf(logFile, "# %s: command %v\n", logPrefix, run.Args)
+
+			cmd := exec.CommandContext(ctx, run.Args[0], run.Args[1:]...)
+			cmd.Dir = volumeDir
+			cmd.Stdout = logFile
+			cmd.Stderr = logFile
+			if err := cmd.Run(); err != nil {
+				fmt.Fprintf(logFile, "# %s: error: %s.\n", logPrefix, err)
+				return nil, errors.Wrap(err, "run command")
+			}
+			fmt.Fprintf(logFile, "# %s: done.\n", logPrefix)
+
+		case "docker":
+			var fromDockerfile string
+			if run.Dockerfile != "" {
+				fromDockerfile = " (built from inline Dockerfile)"
+			}
+			fmt.Fprintf(logFile, "# %s: docker run %v%s\n", logPrefix, run.Image, fromDockerfile)
+
+			cidFile, err := ioutil.TempFile("", prefix+"-container-id")
+			if err != nil {
+				return nil, err
+			}
+			_ = os.Remove(cidFile.Name()) // docker exits if this file exists upon `docker run` starting
+			defer func() {
+				cid, err := ioutil.ReadFile(cidFile.Name())
+				_ = os.Remove(cidFile.Name())
+				if err == nil {
+					ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+					defer cancel()
+					_ = exec.CommandContext(ctx, "docker", "rm", "-f", "--", string(cid)).Run()
+				}
+			}()
+
+			const workDir = "/work"
+			cmd := exec.CommandContext(ctx, "docker", "run",
+				"--rm",
+				"--cidfile", cidFile.Name(),
+				"--workdir", workDir,
+				// TODO!(sqs): on macOS this fails because the volumeDir is /var/folders/... and the
```

Comment:
TODO. Should we use a different location for the temp folder, or since the `cmd.Dir = volumeDir` further down, is it possible to mount the current working directory?